### PR TITLE
feat: enhance AddStory component with MP3 file validation

### PR DESCRIPTION
**Pull request by farmanahmed888**
- Added isMp3File utility function to validate uploaded audio files.
- Implemented error handling for file uploads in AddStory component, displaying appropriate messages for invalid file types.
- Updated file input fields to accept only MP3 files and disable the submit button if there are upload errors.
- Added unit tests for the isMp3File function to ensure correct validation behavior.
- Updated backend routes to reject non-MP3 files when generating SAS tokens for uploads.

### CODE REVIEW

**Review by a4i-architect — COMMENTED** [review]
**High-Level Summary:**

The updates mainly enhance client-side validation for MP3 files, introducing an `isMp3File` utility to verify file extension and MIME type, with corresponding error messages displayed to users and disabling the submit button accordingly. Backend modifications improve robustness by validating `blobName` to ensure only `.mp3` files are processed during SAS token generation and media processing. Test coverage for these validation functions and endpoints has been added, confirming expected behavior.

**Key Issues & Recommendations:**

1. **File Validation Logic:**
   - The frontend validation correctly enforces MP3 restrictions, with user feedback and button disabling.  
   - The utility function properly checks both filename extension and MIME—with the caveat that MIME `"audio/mp3"` is non-standard; most browsers use `"audio/mpeg"` for MP3 files, which is handled correctly.

2. **Error Handling & User Feedback:**
   - Error states (`uploadError`, `answerUploadError`) are effectively shown beneath input fields, preventing form submission while invalid.  
   - Consider resetting error states on successful re-selection for UX clarity.

3. **Backend Validation:**
   - The server-side check for `blobName` adds a necessary security layer.  
   - The URI parsing assumes `blobName` includes the filename with extension; clarify if needed, but generally sufficient.

4. **Code Quality & Consistency:**
   - The front-end code is clean, but the `isMp3File` function could consider accepting more MIME variants if needed (`audio/x-mp3`, etc.).  
   - The test coverage effectively covers validation functions and API responses, though unit tests for `isMp3File` could add more edge cases.

5. **Design & Complexity:**
   - Validation logic is straightforward; no complex or overly complicated patterns detected.  
   - The approach is appropriate: front-end prevents obvious invalid uploads, backend enforces server-side validation.

**Acceptance Score:** 85/100  
**Next steps:** Minor UX improvements, consider broader MIME handling, and ensure validation functions cover edge cases.

File: ContentWebApp/src/utils/fileValidators.js [hunk 1]
```diff
@@ -0,0 +1,7 @@
+export function isMp3File(file) {
+  if (!file) return false;
+  const nameOk = file.name && file.name.toLowerCase().endsWith(".mp3");
+  const typeOk = file.type === "audio/mpeg" || file.type === "audio/mp3";
+  return nameOk || typeOk;
```
Comment (theainerd) [inline ContentWebApp/src/utils/fileValidators.js]: **[P1] `isMp3File` uses OR instead of AND — allows non-MP3 files through**

The `||` (OR) operator means a file only needs to satisfy **one** condition to pass. This creates a security gap:
- A file named `malware.exe` with a spoofed MIME type `audio/mpeg` would pass
- A file named `audio.mp3` with MIME type `application/octet-stream` would also pass

Should be `&&` (AND) to require both `.mp3` extension and valid MIME type. The test "returns true for audio/mpeg MIME type even without .mp3 extension" also validates the wrong behavior and should be updated.

```suggestion
  return nameOk && typeOk;
```

**Review by theainerd — COMMENTED** [review]
**[P1] `isMp3File` uses OR instead of AND — allows non-MP3 files through**

The `||` (OR) operator on `return nameOk || typeOk` means a file only needs to satisfy **one** condition to pass validation. This creates a security gap where:
- A file named `malware.exe` with a spoofed MIME type of `audio/mpeg` would pass
- A file named `audio.mp3` with MIME type `application/octet-stream` would also pass

This should be `&&` (AND) to require **both** the `.mp3` extension and a valid MIME type. The test case "returns true for audio/mpeg MIME type even without .mp3 extension" validates this incorrect behavior and should be updated accordingly.

```suggestion
  return nameOk && typeOk;
```

File: backend-server/src/routes/contentRouter.js [hunk 2]
```diff
@@ -493,10 +493,11 @@ router.get(
   "/sasToken",
   tryCatchWrapper(async (req, res) => {
     const containerName = "input-container";
-    const sasToken = await blobService.getUploadSASToken(
-      req.query.blobName,
-      containerName,
-    );
+    const blobName = req.query.blobName;
+    if (!blobName || !blobName.toLowerCase().endsWith(".mp3")) {
```
Comment (theainerd) [inline backend-server/src/routes/contentRouter.js:497]: **[P2] Backend validation gap**

The `.mp3` check here is good, but the same validation is missing on the `POST /content` endpoint. If someone bypasses the frontend and calls the API directly, they can save content referencing non-MP3 files (e.g., `.wav`, `.aac`). The `processAudioContent.js` job then silently skips non-MP3 items, leaving content records with unprocessed audio URLs.

Consider also validating audio URLs in the `POST /content` handler.

**Review by theainerd — COMMENTED** [review]
**[P2] Backend validation gap — `POST /content` route is unprotected**

The `/sasToken` endpoint (line 497) now correctly rejects non-MP3 uploads. However, the `POST /content` endpoint saves content directly without validating that `audioContent[].audioUrl` references an MP3 file. If someone bypasses the frontend and calls the API directly, they could save content referencing non-MP3 files. The processing job in `processAudioContent.js` silently skips non-MP3 items, leaving content records with unprocessed audio URLs and no user notification.

Consider adding MP3 validation in the `POST /content` handler as well:
```js
for (const item of req.body.audioContent || []) {
  if (item.audioUrl && !item.audioUrl.toLowerCase().endsWith('.mp3')) {
    return res.status(400).json({ error: 'Only .mp3 audio files are allowed.' });
  }
}
```

**Review by a4i-architect — COMMENTED** [review]
**Summary of Key Changes & Recommendations:**

1. **Validation for MP3 Files**:  
   - The new `isMp3File` utility enforces extension and MIME type checks, which enhances robustness.
   - Proper error messages (`uploadError`/`answerUploadError`) ensure UX clarity.
   - The validation logic appropriately prevents non-mp3 uploads before form submission.

2. **Code Quality & Redundancy**:  
   - Redundant comments and commented-out code in handlers can be cleaned up.
   - Using both filename extension and MIME type checks is good, but MIME type detection from `File` can sometimes be unreliable; consider just extension check if consistent.
   - Duplicate/error-prone logic in `handleUploadFile` and `handleUploadAnswerFile` could be abstracted into a shared helper for DRYness.

3. **UI/UX Handling**:
   - Error messages appear conditionally, improving user feedback.
   - Submission button disables when errors are present, which is good, but should also consider disabling during ongoing upload/processing.

4. **Backend Validation & Processing**:
   - The backend validation of `.mp3` extension enhances security and data integrity.
   - The `processAudioContent` function has been updated to skip non-mp3 files, avoiding processing errors.
   - Logging of audio URLs and error handling during processing suggests improved resilience.

5. **Missing or Ill-Typed Areas**:
   - `isMp3File` checks are correct but could be tightened for edge cases.
   - The optional chaining (`?.`) in some parts could enhance safety.
   - The `audioUrl` validation on upload and in backend checks is consistent; ensure they stay synchronized.

6. **Design & Complexity**:
   - Overall, the design pattern is straightforward; validation exists early, reducing bad data flow.
   - Complexity is moderate; consider centralizing validation and error handling logic for maintainability.

**Score**: 85/100  
- Strengths in validation and error handling, slight room for code DRYness, and further UX polish.

---

**Next steps**:
- Refactor upload handlers to reduce duplication.
- Ensure consistent validation logic across frontend/backend.
- Further improve user feedback (e.g., loading states).
- Confirm test coverage includes all edge cases, especially for file validation.

File: backend-server/src/routes/contentRouter.js (hunk 2 above)
Comment (farmanahmed888) [reply backend-server/src/routes/contentRouter.js:497]: addressed in latest revision

File: ContentWebApp/src/utils/fileValidators.js (hunk 1 above)
Comment (farmanahmed888) [reply ContentWebApp/src/utils/fileValidators.js]: addressed in latest revision

File: backend-server/src/routes/contentRouter.js [hunk 3]
```diff
@@ -573,6 +574,13 @@ router.post(
   tryCatchWrapper(async (req, res) => {
     let content = new ContentV3(req.body);
     content.creation_time = Math.floor(Date.now() / 1000);
+    console.log(content.audioContent);
```
Comment (N1kunj1998) [inline backend-server/src/routes/contentRouter.js]: this line will pollute the production log stream with large arrays on every content save, we should remove it.

File: backend-server/src/routes/contentRouter.js [hunk 4]
```diff
@@ -573,6 +574,13 @@ router.post(
...
+    // Validate audioContent entries to ensure uploaded audio URLs reference .mp3 files.
+    for (const item of content.audioContent || []) {
+      if (item.audioUrl && !item.audioUrl.toLowerCase().endsWith(".mp3")) {
+        return res.status(400).json({ error: "Only .mp3 audio files are allowed." });
```
Comment (N1kunj1998) [inline backend-server/src/routes/contentRouter.js:581]: we should keep error message same everywhere for consistency, as in same file line 498 has different error message(missing "audio" word).

File: backend-server/src/jobs/processAudioContent.js [hunk 5]
```diff
@@ -97,10 +97,22 @@ async function processNewContent(job) {
...
+          console.error(`Skipping non-mp3 file during processing: ${ip_url}`);
+          continue;
+        }
+
+        audioContentItem.audioUrl = await generateWAVFileAndUploadToOutputContainer(
+          ip_url,
+          contentDoc.id
+        );
+      } catch (err) {
+        console.error(`Failed to process audio content item ${ip_url}:`, err);
```
Comment (N1kunj1998) [inline backend-server/src/jobs/processAudioContent.js:113]: If processing fails still we are saving audio and not giving any error, rather we should propogate error.

```
} catch (err) {
  console.error(`Failed to process audio content item ${ip_url}:`, err);
  throw err;  // propagate — outer catch will mark job as failed
}
```

File: ContentWebApp/tests/utils/fileValidators.test.js [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+  test("returns false for non-mp3 extension and non-mpeg MIME type", () => {
+    const mockFile = { name: "audio.wav", type: "audio/wav" };
+    expect(isMp3File(mockFile)).toBe(false);
+  });
+
+  test("returns false for null or undefined file", () => {
+    expect(isMp3File(null)).toBe(false);
+    expect(isMp3File(undefined)).toBe(false);
+  });
```
Comment (N1kunj1998) [inline ContentWebApp/tests/utils/fileValidators.test.js:27]: We should add this one more test case which covers scenario of user renaming a .wav file to .mp3

```
test("returns false for .mp3 filename with non-mpeg MIME type", () => {
  const mockFile = { name: "audio.mp3", type: "audio/wav" };
  expect(isMp3File(mockFile)).toBe(false);
});
```

**Comment by N1kunj1998** [comment]
I have checked functionality it is working fine but we need to make some modifications as added in comments.

**Review by a4i-architect — COMMENTED** [review]
### High-Level Review Summary

The refactor introduces client-side validation for MP3 files, utilizing a utility function `isMp3File`, and displays error messages, preventing users from uploading invalid files. On the backend, validation is added to ensure only MP3 files are processed and stored, with appropriate HTTP status responses. The tests are updated for the new validation logic.

### Key Issues & Recommendations

1. **Utility function (`isMp3File`)**:
   - Correctly encapsulates validation but could also check MIME types more robustly (e.g., allow more MIME variants).
   - Already covered by tests, which is good.

2. **Client-side validation**:
   - Properly prevents invalid files from being uploaded, with user feedback.
   - Disabled submit button logic considers error states, which improves UX.

3. **Backend validation**:
   - Ensures only MP3 files are accepted during upload URL generation and content saving.
   - The URL parsing logic in `processAudioContent.js` is sound, but consider adding more robust validation for URL correctness.

4. **Potential issues**:
   - `isMp3File` is strict about both filename extension and MIME type. While good, some valid MP3 files may have different MIME types; consider relaxing or documenting.
   - Front-end error state resets are handled well.
   - Disabling submit button on upload errors prevents invalid submission; ensure this logic aligns with all possible error states.

5. **Additional suggestions**:
   - In the backend, after validation, another layer such as schema validation could reinforce data integrity.
   - For the server-side URL parsing, consider try-catch to handle malformed URLs gracefully.
   - Remove commented-out code or legacy code snippets for clarity.
   - Think about unifying validation logic: maybe create a shared validation package or middleware for comprehensive validation.

### Overall Score: **85/100**

**Acceptance recommended** after minor enhancements for URL validation robustness and clearer documentation of validation logic.

File: ContentWebApp/tests/utils/fileValidators.test.js (hunk 6 above)
Comment (farmanahmed888) [reply ContentWebApp/tests/utils/fileValidators.test.js:27]: done

File: backend-server/src/jobs/processAudioContent.js (hunk 5 above)
Comment (farmanahmed888) [reply backend-server/src/jobs/processAudioContent.js:113]: done

File: backend-server/src/routes/contentRouter.js (hunk 4 above)
Comment (farmanahmed888) [reply backend-server/src/routes/contentRouter.js:581]: done

File: backend-server/src/routes/contentRouter.js (hunk 3 above)
Comment (farmanahmed888) [reply backend-server/src/routes/contentRouter.js]: done

**Review by N1kunj1998 — APPROVED** [review]
Tested on Local working fine, quite extensive corner cases are covered.

**Review by theainerd — APPROVED** [review]
The patch correctly implements MP3 file validation across frontend and backend. All critical review comments from previous rounds (P1 OR→AND fix, backend POST validation gap, error propagation, test coverage) have been addressed.
